server/server.go

```diff
@@ -1930,9 +1952,34 @@ func (s *Server) createClient(conn net.Conn, ws *websocket) *client {
 	// Re-Grab lock
 	c.mu.Lock()
 
+	tlsRequired := ws == nil && info.TLSRequired
+	var pre []byte
+	// If we have both TLS and non-TLS allowed we need to see which
+	// one the client wants.
+	if opts.TLSConfig != nil && opts.AllowNonTLS {
+		pre = make([]byte, 8)
+		c.nc.SetReadDeadline(time.Now().Add(25 * time.Millisecond))
+		n, err := c.nc.Read(pre[:])
+		c.nc.SetReadDeadline(time.Time{})
+		pre = pre[:n]
+		// Assume TLS unless we see nothing or CONNECT.
+		if err != nil || bytes.Contains(pre, []byte("CONNECT")) {
```

Comment:
I disagree with that. Creating a buffer of 8 bytes and calling Read() with this buffer does not guarantee that you are going to receive 8 bytes. And don't think this is not possible. I have fixed an issue similar to that when working in a previous company and the number of expected bytes with even smaller.

Also, if err is not nil, it could be because of deadline, why does that mean that it is going to be non TLS? The client creates TCP connection, this server sends INFO, client processes and then send CONNECT. If client is far apart, it may take more than the deadline, regardless if client is going to do TLS or not.

Reply:
I could do a ReadAll. Timeout == telnet or no CONNECT.

Reply:
Agree timeout may need some tweaking.

Reply:
Yeah, timeout will be tough, because if client is far, even if it is going to do TLS or CONNECT, it may take more than that.. And keep in mind that this is already a client go routine, so maybe if we were not doing that under client lock, it would be ok to wait for longer, no?